README.md
# `embedded-picofont` [![Star me](https://img.shields.io/github/stars/althonos/embedded-picofont.svg?style=social&label=Star&maxAge=3600)](https://github.com/althonos/embedded-picofont/stargazers)

*The PICO-8 font to use with [`embedded-graphics`](https://crates.io/crates/embedded-graphics).*

[![TravisCI](https://img.shields.io/travis/com/althonos/embedded-picofont/master.svg?maxAge=600&style=flat-square)](https://travis-ci.com/althonos/embedded-picofont/branches)
[![Codecov](https://img.shields.io/codecov/c/gh/althonos/embedded-picofont/master.svg?style=flat-square&maxAge=600)](https://codecov.io/gh/althonos/embedded-picofont)
[![License](https://img.shields.io/badge/license-MIT-blue.svg?style=flat-square&maxAge=2678400)](https://choosealicense.com/licenses/mit/)
[![Source](https://img.shields.io/badge/source-GitHub-303030.svg?maxAge=2678400&style=flat-square)](https://github.com/althonos/embedded-picofont)
[![Crate](https://img.shields.io/crates/v/embedded-picofont.svg?maxAge=600&style=flat-square)](https://crates.io/crates/embedded-picofont)
[![Documentation](https://img.shields.io/badge/docs.rs-latest-4d76ae.svg?maxAge=2678400&style=flat-square)](https://docs.rs/embedded-picofont)
[![Changelog](https://img.shields.io/badge/keep%20a-changelog-8A0707.svg?maxAge=2678400&style=flat-square)](https://github.com/althonos/embedded-picofont/blob/master/CHANGELOG.md)
[![GitHub issues](https://img.shields.io/github/issues/althonos/embedded-picofont.svg?style=flat-square)](https://github.com/althonos/embedded-picofont/issues)

## Overview

This crate provides the super tiny 4x6 font of the PICO-8 fantasy console as
an [`embedded_graphics::fonts::Font`](https://docs.rs/embedded-graphics/0.6.2/embedded_graphics/fonts/trait.Font.html):

![Font](https://www.lexaloffle.com/gfx/pico8_font.png)

Please note the PICO-8 itself only use the uppercase characters, as the lowercase
chars can get *really* hard to read if the display is not upscaled. As such, it
is advised to only use this font to display uppercase characters.

## Usage

Use [`TextStyle`](https://docs.rs/embedded-graphics/0.6.2/embedded_graphics/style/struct.TextStyle.html)
to attach the PICO-8 font to a text:
```rust
use embedded_picofont::FontPico;
let text = Text::new("Hello world!", Point::new(0, 0))
    .into_styled(TextStyle::new(FontPico, Gray8::WHITE));
```
![Hello world](https://github.com/althonos/embedded-picofont/raw/master/static/helloworld.png)

The PICO-8 also has wide characters: these can be displayed using two smaller
characters in the `128..255` char range:
```rust
use embedded_picofont::{text_pico, FontPico};
let text = Text::new("PRESS \u{96}\u{97} TO GO BACK", Point::new(0, 0))
    .into_styled(TextStyle::new(FontPico, Gray8::WHITE));
```
![Press left to go back](https://github.com/althonos/embedded-picofont/raw/master/static/goback.png)

See the [`embedded-graphics` documentation](https://docs.rs/embedded-graphics/)
for more information on how to use this crate.

## License

* The [original PICO-8 font](https://www.lexaloffle.com/pico-8.php?page=faq)
  is available under the [CC-0 license](https://creativecommons.org/share-your-work/public-domain/cc0/).
* This crate is released under the [MIT License](https://opensource.org/licenses/mit-license.php).


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).


## [Unreleased]

[Unreleased]: https://github.com/althonos/embedded-picofont/compare/v0.2.1...HEAD

### Changed
- **(breaking)** Bumped `embedded-graphics` dependency to `v0.7.0`.
- **(breaking)** The `FontPico` struct is now a `const` called `PICO_FONT`.

## [0.2.1] - 2020-11-27

[0.2.1]: https://github.com/althonos/embedded-picofont/compare/v0.2.0...v0.2.1

### Fixed
- Broken link to the GitHub page in `Cargo.toml` manifest.

### Changed
- Bumped `lodepng` build dependency to `v3.2`.
- Bumped `bitvec` build dependency to `v0.19.4`.


## [0.2.0] - 2020-04-18

[0.2.0]: https://github.com/althonos/embedded-picofont/compare/v0.1.1...v0.2.0

### Changed
- Bumped `embedded-graphics` dependency to `v0.6.0`.
### Removed
- `text_pico` macro (not useful with `embedded-graphics` new API).


## [0.1.1] - 2019-10-30

[0.1.1]: https://github.com/althonos/embedded-picofont/compare/v0.1.0...v0.1.1

### Changed
- Bumped `bitvec` build dependency to `v0.15.0`.


## [0.1.0] - 2019-07-02

[0.1.0]: https://github.com/althonos/embedded-picofont/compare/2ae49c2...v0.1.0

Initial release.


tests/draw.rs
extern crate embedded_graphics;
extern crate embedded_picofont;
extern crate lodepng;

use std::path::Path;
use std::path::PathBuf;

use embedded_graphics::mono_font::MonoTextStyle;
use embedded_graphics::pixelcolor::Gray8;
use embedded_graphics::prelude::*;
use embedded_graphics::text::Text;
use embedded_picofont::PICO_FONT;
use lodepng::ffi::ColorType;
use lodepng::Bitmap;
use lodepng::Grey;
use lodepng::Image;

pub struct GreyDisplay(Bitmap<Grey<u8>>);

impl DrawTarget for GreyDisplay {
    type Color = Gray8;
    type Error = core::convert::Infallible;

    fn draw_iter<I>(&mut self, pixels: I) -> Result<(), Self::Error>
    where
        I: IntoIterator<Item = Pixel<Self::Color>>,
    {
        for Pixel(coord, color) in pixels.into_iter() {
            if (coord[0] as usize) < self.0.width && (coord[1] as usize) < self.0.height {
                let index = coord[0] as usize + coord[1] as usize * self.0.width;
                self.0.buffer[index] = Grey(color.luma());
            }
        }

        Ok(())
    }
}

impl OriginDimensions for GreyDisplay {
    fn size(&self) -> Size {
        Size::new(self.0.width as u32, self.0.height as u32)
    }
}

fn test_grayscale(reference: &Path, text: &str) {
    let exp = match lodepng::decode_file(reference, ColorType::GREY, 8) {
        Ok(Image::Grey(bmp)) => bmp,
        Ok(_other) => panic!("invalid PNG format"),
        Err(e) => panic!("could not load reference image: {}", e),
    };

    let mut display: GreyDisplay = GreyDisplay(Bitmap {
        buffer: vec![Grey(0); exp.width * exp.height],
        width: exp.width,
        height: exp.height,
    });

    Text::new(
        text,
        Point::new(0, 6),
        MonoTextStyle::new(&PICO_FONT, Gray8::WHITE),
    )
    .draw(&mut display)
    .unwrap();

    if exp.buffer != display.0.buffer {
        lodepng::encode_file(
            "/tmp/expected.png",
            &exp.buffer,
            exp.width,
            exp.height,
            ColorType::GREY,
            8,
        )
        .unwrap();
        lodepng::encode_file(
            "/tmp/actual.png",
            &display.0.buffer,
            display.0.width,
            display.0.height,
            ColorType::GREY,
            8,
        )
        .unwrap();
        panic!("images don't match: see /tmp");
    }
}

#[test]
fn test_hello_world() {
    test_grayscale(&PathBuf::from("tests/helloworld.png"), "Hello, world!");
}

#[test]
fn test_abcdefg() {
    test_grayscale(&PathBuf::from("tests/abcdefg.png"), "a.b,c:d;e'f!g");
}

#[test]
fn test_buttons() {
    test_grayscale(
        &PathBuf::from("tests/buttons.png"),
        "\u{96}\u{97}\u{86}\u{87}\u{a2}\u{a3}\u{a8}\u{a9}\u{9c}\u{9d}\u{ae}\u{af}",
    );
}

#[test]
fn test_unicode() {
    test_grayscale(&PathBuf::from("tests/unicode.png"), "😀");
}


src/build.rs
// extern crate image;
extern crate bitvec;
extern crate lodepng;

use bitvec::order::Msb0;
use lodepng::ffi::ColorType;
use lodepng::Image;
use std::env::var;
use std::io::Write;
use std::path::PathBuf;

fn main() {
    // Load the original image.
    let mut dst: bitvec::vec::BitVec<Msb0, u8> = bitvec::vec::BitVec::new();
    let bmp = match lodepng::decode_file("src/font.png", ColorType::GREY, 8) {
        Ok(Image::Grey(bmp)) => bmp,
        Ok(_) => panic!("unexpected image format"),
        Err(e) => panic!("PNG loading error: {}", e),
    };

    // Create a bitmap of depth 1
    for pixel in bmp.buffer.iter() {
        dst.push(pixel.0 != 0);
    }

    // Write the generated bitmap.
    let filename = PathBuf::from(var("OUT_DIR").unwrap()).join("font.raw");
    std::fs::File::create(filename)
        .and_then(|mut f| f.write_all(&dst.into_vec()))
        .expect("could ");
}


src/lib.rs
//! *The PICO-8 font to use with [`embedded-graphics`](https://crates.io/crates/embedded-graphics).*
//!
//! [![TravisCI](https://img.shields.io/travis/althonos/embedded-picofont/master.svg?maxAge=600&style=flat-square)](https://travis-ci.org/althonos/embedded-picofont/branches)
//! [![Codecov](https://img.shields.io/codecov/c/gh/althonos/embedded-picofont/master.svg?style=flat-square&maxAge=600)](https://codecov.io/gh/althonos/embedded-picofont)
//! [![License](https://img.shields.io/badge/license-MIT-blue.svg?style=flat-square&maxAge=2678400)](https://choosealicense.com/licenses/mit/)
//! [![Source](https://img.shields.io/badge/source-GitHub-303030.svg?maxAge=2678400&style=flat-square)](https://github.com/althonos/embedded-picofont)
//! [![Crate](https://img.shields.io/crates/v/embedded-picofont.svg?maxAge=600&style=flat-square)](https://crates.io/crates/embedded-picofont)
//! [![Documentation](https://img.shields.io/badge/docs.rs-latest-4d76ae.svg?maxAge=2678400&style=flat-square)](https://docs.rs/embedded-picofont)
//! [![Changelog](https://img.shields.io/badge/keep%20a-changelog-8A0707.svg?maxAge=2678400&style=flat-square)](https://github.com/althonos/embedded-picofont/blob/master/CHANGELOG.md)
//! [![GitHub issues](https://img.shields.io/github/issues/althonos/embedded-picofont.svg?style=flat-square)](https://github.com/althonos/embedded-picofont/issues)
//!
//! # Overview
//!
//! This crate provides the super tiny 4x6 font of the PICO-8 fantasy console as an [`embedded_graphics::fonts::Font`](https://docs.rs/embedded-graphics/0.6.2/embedded_graphics/fonts/trait.Font.html):
//!
//! <img src="https://www.lexaloffle.com/gfx/pico8_font.png" style="width:100%;image-rendering:-moz-crisp-edges;image-rendering:-o-crisp-edges;image-rendering:-webkit-optimize-contrast;image-rendering: crisp-edges;-ms-interpolation-mode: nearest-neighbor;">
//!
//! Please note the PICO-8 itself only use the uppercase characters, as the lowercase
//! chars can get *really* hard to read if the display is not upscaled. As such, it
//! is advised to only use this font to display uppercase characters.
//!
//! # Usage
//!
//! Use [`MonoTextStyle`](https://docs.rs/embedded-graphics/0.7.0/embedded_graphics/mono_font/struct.MonoTextStyle.html)
//! to attach the PICO-8 font to a text:
//! ```rust
//! use embedded_graphics::prelude::*;
//! use embedded_graphics::text::Text;
//! use embedded_graphics::pixelcolor::Gray8;
//! use embedded_graphics::mono_font::MonoTextStyle;
//! use embedded_picofont::PICO_FONT;
//!
//! let style = MonoTextStyle::new(&PICO_FONT, Gray8::WHITE);
//!
//! let text = Text::new("Hello world!", Point::new(0, 6), style);
//! ```
//! ![Hello world](https://github.com/althonos/embedded-picofont/raw/master/static/helloworld.png)
//!
//! The PICO-8 also has wide characters: these can be displayed using two smaller
//! characters in the `128..255` char range:
//! ```rust
//! # use embedded_graphics::prelude::*;
//! # use embedded_graphics::text::Text;
//! # use embedded_graphics::pixelcolor::Gray8;
//! # use embedded_graphics::mono_font::MonoTextStyle;
//! # use embedded_picofont::PICO_FONT;
//!
//! let style = MonoTextStyle::new(&PICO_FONT, Gray8::WHITE);
//!
//! let text = Text::new("PRESS \u{96}\u{97} TO GO BACK", Point::new(0, 6), style);
//! ```
//! ![Press left to go back](https://github.com/althonos/embedded-picofont/raw/master/static/goback.png)
//!
//! See the [`embedded-graphics` documentation](https://docs.rs/embedded-graphics/)
//! for more information on how to use this crate.
//!
//! # License
//!
//! * The [original PICO-8 font](https://www.lexaloffle.com/pico-8.php?page=faq)
//!   is available under the [CC-0 license](https://creativecommons.org/share-your-work/public-domain/cc0/).
//! * This crate is released under the [MIT License](https://opensource.org/licenses/mit-license.php).
//!

#![no_std]

use embedded_graphics::geometry::Size;
use embedded_graphics::image::ImageRaw;
use embedded_graphics::mono_font::DecorationDimensions;
use embedded_graphics::mono_font::MonoFont;

/// The 4x6 pixel monospace font used by the PICO-fantasy 8 console.
///
/// # Examples
///
/// ## Write some text to the screen at the default `(0, 0)` position
///
/// ```rust
/// # use embedded_graphics::prelude::*;
/// # use embedded_graphics::text::Text;
/// # use embedded_graphics::pixelcolor::Gray8;
/// # use embedded_graphics::mono_font::MonoTextStyle;
/// # use embedded_picofont::PICO_FONT;
/// # use embedded_graphics::mock_display::MockDisplay;
/// # let mut display = MockDisplay::default();
/// let style = MonoTextStyle::new(&PICO_FONT, Gray8::WHITE);
///
/// Text::new("Hello Rust!", Point::new(0, 6), style)
///     .draw(&mut display);
/// ```
///
/// ## Translate text by (20px, 30px)
///
/// ```rust
/// # use embedded_graphics::prelude::*;
/// # use embedded_graphics::text::Text;
/// # use embedded_graphics::pixelcolor::Gray8;
/// # use embedded_graphics::mono_font::MonoTextStyle;
/// # use embedded_picofont::PICO_FONT;
/// # use embedded_graphics::mock_display::MockDisplay;
/// # let mut display = MockDisplay::default();
/// let style = MonoTextStyle::new(&PICO_FONT, Gray8::WHITE);
///
/// Text::new("Hello Rust!", Point::new(20, 30), style)
///     .draw(&mut display);
/// ```
///
/// ## Add some styling to the text
///
/// Use the [`MonoTextStyleBuilder`](https://docs.rs/embedded-graphics/0.7.0/embedded_graphics/mono_font/struct.MonoTextStyleBuilder.html)
/// trait to edit the colors of the rendered text:
///
/// ```rust
/// # use embedded_graphics::prelude::*;
/// # use embedded_graphics::text::Text;
/// # use embedded_graphics::pixelcolor::Rgb888;
/// # use embedded_graphics::mono_font::MonoTextStyleBuilder;
/// # use embedded_picofont::PICO_FONT;
/// # use embedded_graphics::mock_display::MockDisplay;
/// # let mut display = MockDisplay::default();
/// let style = MonoTextStyleBuilder::new()
///     .font(&PICO_FONT)
///     .text_color(Rgb888::MAGENTA)
///     .background_color(Rgb888::BLACK)
///     .build();
///
/// Text::new("Hello Rust!", Point::new(0, 6), style)
///     .draw(&mut display);
/// ```
pub const PICO_FONT: MonoFont = MonoFont {
    image: ImageRaw::new_binary(include_bytes!(concat!(env!("OUT_DIR"), "/font.raw")), 128),
    glyph_mapping: &char_offset,
    character_size: Size::new(4, 6),
    character_spacing: 0,
    baseline: 6,
    underline: DecorationDimensions::new(6, 1),
    strikethrough: DecorationDimensions::default_strikethrough(6),
};

fn char_offset(c: char) -> usize {
    if c <= '\u{7f}' {
        return (c as usize) * 2;
    } else if c <= '\u{b3}' {
        return c as usize + 0x80;
    }
    ('?' as usize) * 2
}
